pkg/cluster/mutate_create.go

```diff
@@ -0,0 +1,125 @@
+)
+
+type CreateMutator struct {
+	baseDomain string
+	logger     micrologger.Logger
+}
+
+type CreateMutatorConfig struct {
+	BaseDomain string
+	Logger     micrologger.Logger
+}
+
+func NewCreateMutator(config CreateMutatorConfig) (*CreateMutator, error) {
+	if config.BaseDomain == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.BaseDomain must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	v := &CreateMutator{
+		baseDomain: config.BaseDomain,
+		logger:     config.Logger,
+	}
+
+	return v, nil
+}
+
+func (m *CreateMutator) Mutate(ctx context.Context, request *v1beta1.AdmissionRequest) ([]mutator.PatchOperation, error) {
+	var result []mutator.PatchOperation
+
+	if request.DryRun != nil && *request.DryRun {
+		m.logger.LogCtx(ctx, "level", "debug", "message", "Dry run is not supported. Request processing stopped.")
+		return result, nil
+	}
+
+	clusterCR := &capiv1alpha3.Cluster{}
+	if _, _, err := mutator.Deserializer.Decode(request.Object.Raw, nil, clusterCR); err != nil {
+		return []mutator.PatchOperation{}, microerror.Maskf(errors.ParsingFailedError, "unable to parse Cluster CR: %v", err)
+	}
+
+	patch, err := m.ensureClusterNetwork(ctx, clusterCR)
+	if err != nil {
+		return []mutator.PatchOperation{}, microerror.Mask(err)
+	}
+	if patch != nil {
+		result = append(result, *patch)
+	}
+
+	patch, err = m.ensureControlPlaneEndpointHost(ctx, clusterCR)
+	if err != nil {
+		return []mutator.PatchOperation{}, microerror.Mask(err)
+	}
+	if patch != nil {
+		result = append(result, *patch)
+	}
+
+	patch, err = m.ensureControlPlaneEndpointPort(ctx, clusterCR)
+	if err != nil {
+		return []mutator.PatchOperation{}, microerror.Mask(err)
+	}
+	if patch != nil {
+		result = append(result, *patch)
+	}
+
+	return result, nil
+}
+
+func (m *CreateMutator) Log(keyVals ...interface{}) {
+	m.logger.Log(keyVals...)
+}
+
+func (m *CreateMutator) Resource() string {
+	return "cluster"
+}
+
+func (m *CreateMutator) ensureClusterNetwork(ctx context.Context, clusterCR *capiv1alpha3.Cluster) (*mutator.PatchOperation, error) {
+	// Ensure ClusterNetwork is set.
+	if clusterCR.Spec.ClusterNetwork == nil {
+		clusterNetwork := capiv1alpha3.ClusterNetwork{
+			APIServerPort: to.Int32Ptr(key.ControlPlaneEndpointPort),
+			ServiceDomain: key.ServiceDomain(clusterCR.Name, m.baseDomain),
+			Services: &capiv1alpha3.NetworkRanges{
+				CIDRBlocks: []string{
+					key.ClusterNetworkServiceCIDR,
+				},
+			},
+		}
+
+		return mutator.PatchAdd("/spec/clusterNetwork", clusterNetwork), nil
+	}
+
+	return nil, nil
+}
+
+func (m *CreateMutator) ensureControlPlaneEndpointHost(ctx context.Context, clusterCR *capiv1alpha3.Cluster) (*mutator.PatchOperation, error) {
+	if clusterCR.Spec.ControlPlaneEndpoint.Host == "" {
+		return mutator.PatchAdd("/spec/controlPlaneEndpoint/host", fmt.Sprintf("api.%s.%s", clusterCR.Name, m.baseDomain)), nil
```

Comment:
should `fmt.Sprintf("api.%s.%s", clusterCR.Name, m.baseDomain)` be returned by a function in the `key` package like the rest of the values? WDYT?

Reply:
yeah totally